aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,764 @@
+        }
+    }
+
+    fn get_outstanding_requests_name(prefix: &str) -> String {
+        [prefix, Self::OUTSTANDING_REQUESTS_METRIC_NAME]
+            .into_iter()
+            .join("_")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        sync::{
+            atomic::{AtomicU32, Ordering},
+            Arc,
+        },
+        time::Duration,
+    };
+
+    use async_trait::async_trait;
+    use backoff::{future::retry, ExponentialBackoff};
+    use futures::{future::join_all, Future};
+    use janus_aggregator_core::test_util::noop_meter;
+    use janus_core::test_util::install_test_trace_subscriber;
+    use opentelemetry_sdk::metrics::data::Gauge;
+    use quickcheck::{quickcheck, Arbitrary, TestResult};
+    use tokio::{
+        runtime::Builder as RuntimeBuilder,
+        sync::Notify,
+        task::{yield_now, JoinHandle},
+        time::timeout,
+    };
+    use tracing::debug;
+    use trillium::{Conn, Handler, Status};
+    use trillium_testing::{assert_ok, assert_status, methods::get};
+
+    use crate::{
+        aggregator::queue::{queued_lifo, LIFORequestQueue, Metrics},
+        metrics::test_util::InMemoryMetricsInfrastructure,
+    };
+
+    /// Some tests busy loop waiting for a condition to become true. Avoid hanging broken tests
+    /// indefinitely by wrapping those tests in a timeout.
+    const TEST_TIMEOUT: Duration = Duration::from_secs(15);
+
+    async fn get_outstanding_requests_gauge(
+        metrics: &InMemoryMetricsInfrastructure,
+        meter_prefix: &str,
+    ) -> Option<usize> {
+        Some(
+            metrics
+                .collect()
+                .await
+                // The metric may not be immediately available when we need it, so return an Option
+                // instead of unwrapping.
+                .get(&Metrics::get_outstanding_requests_name(meter_prefix))?
+                .data
+                .as_any()
+                .downcast_ref::<Gauge<u64>>()
+                .unwrap()
+                .data_points[0]
+                .value
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    struct HangingHandler {
+        unhang: Arc<Notify>,
+    }
+
+    #[async_trait]
+    impl Handler for HangingHandler {
+        async fn run(&self, conn: trillium::Conn) -> trillium::Conn {
+            let _ = self.unhang.notified().await;
+            conn.ok("hello")
+        }
+    }
+
+    async fn fill_queue(
+        handler: Arc<impl Handler>,
+        concurrency: u32,
+        depth: usize,
+        metrics: &InMemoryMetricsInfrastructure,
+        meter_prefix: &str,
+    ) -> Vec<JoinHandle<()>> {
+        debug!("filling queue");
+
+        let mut requests = Vec::new();
+
+        let backoff = ExponentialBackoff {
+            initial_interval: Duration::from_nanos(1),
+            max_interval: Duration::from_nanos(30),
+            multiplier: 2.0,
+            ..Default::default()
+        };
+
+        debug!("spawning requests");
```

Comment:
I have a moderate preference for keeping them. Since these tests have the propensity to hang and timeout on failure, rerunning the test with `RUST_LOG=debug cargo test foo -- --nocapture` is helpful to see where the test is hanging.

I'm not sure of the binary size/runtime cost/compile time impact is though (i.e. the impact to CI).

Reply:
OK; I won't hold the PR up over this, especially if we're concerned about regressions resulting in deadlocks (which seems reasonable).